packages/runtime/src/helpers/functions.ts

```diff
@@ -196,6 +216,125 @@ export const setupImageFunction = async ({
   }
 }
 
+const traceRequiredServerFiles = async (publish: string): Promise<string[]> => {
+  const requiredServerFilesPath = join(publish, 'required-server-files.json')
+  const { files } = (await readJSON(requiredServerFilesPath)) as { files: string[] }
+  files.push(requiredServerFilesPath)
+  return files
+}
+
+const traceNextServer = async (publish: string, baseDir: string): Promise<string[]> => {
+  const nextServerDeps = await getDependenciesOfFile(join(publish, 'next-server.js'))
+
+  // this takes quite a long while, but it's only done once per build.
+  // theoretically, it's only needed once per version of Next.js 🤷
+  const { fileList } = await nodeFileTrace(nextServerDeps, { base: '/' })
```

Comment:
I wonder whether we could persist the result of this between builds with the build cache?

Reply:
I've dabbled around with a barebones Next.js repo, and it looks like `next-server.js.nft.json` already contains all we need there. That was somehow not the case in the Demo repo here, so I added the `nodeFileTrace` call here - ideally, we wouldn't need this at all 🤔 I'll see if we can remove this. Then we don't need to cache it :D